drivers/mfd/tssupervisor-core.c

```diff
@@ -119,6 +121,66 @@ static ssize_t wake_en_show(struct device *dev,
 }
 static DEVICE_ATTR_RW(wake_en);
 
+static ssize_t serial_store(struct device *dev, struct device_attribute *attr,
+			      const char *buf, size_t count)
+{
+	struct ts_supervisor *super = dev_get_drvdata(dev);
+	unsigned int ctrl = 0;
+	u8 sn[6];
+	int ret;
+
+	ret = regmap_read(super->regmap, SUPER_SERIAL_CTRL, &ctrl);
+	if (ret)
+		return ret;
+
+	/* Dont write the SN if its already written */
+	if (ctrl & SUPER_SN_LOCKED)
+		return -EEXIST;
+
+	if (count < 17)
+		return -EIO;
+
+	if (sscanf(buf, "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+		   &sn[0], &sn[1], &sn[2], &sn[3], &sn[4], &sn[5]) != 6)
+		return -EINVAL;
+
+	ret = regmap_bulk_write(super->regmap, SUPER_SERIAL0, sn, 3);
+	if (ret)
+		return ret;
+
+	ret = regmap_update_bits(super->regmap, SUPER_SERIAL_CTRL,
+				 SUPER_SN_LOCKED,
+				 SUPER_SN_LOCKED);
+
+	return ret ? ret : count;
+}
+
+static ssize_t serial_show(struct device *dev,
+				struct device_attribute *attr, char *buf)
+{
+	struct ts_supervisor *super = dev_get_drvdata(dev);
+	unsigned int ctrl;
+	u8 sn[6];
+	int ret;
+
+	ret = regmap_read(super->regmap, SUPER_SERIAL_CTRL, &ctrl);
+	if (ret)
+		return ret;
+
+	/* Dont show the serial number if its not written yet */
+	if (!(ctrl & SUPER_SN_LOCKED))
+		return -EINVAL;
+
+	ret = regmap_bulk_read(super->regmap, SUPER_SERIAL0, sn, 3);
+	if (ret)
+		return ret;
+	ret = sprintf(buf, "%02x:%02x:%02x:%02x:%02x:%02x\n",
```

Comment:
Same with above. Is buf guaranteed to be any specific size? We may want to check its length before writing to it if not. Could also use snprintf on buf, but I don't think thats hugely necessary if we do length check/are guaranteed buf size.